server/sonar-server/src/main/java/org/sonar/server/rule/CachingRuleFinder.java

```diff
@@ -0,0 +1,196 @@
+      return ImmutableListMultimap.of();
+    }
+    return dbClient.ruleDao().selectRuleParamsByRuleKeys(dbSession, ruleKeys)
+      .stream()
+      .collect(MoreCollectors.index(RuleParamDto::getRuleId));
+  }
+
+  @Override
+  @Deprecated
+  @CheckForNull
+  public Rule findById(int ruleId) {
+    return rulesById.get(ruleId);
+  }
+
+  @Override
+  @CheckForNull
+  public Rule findByKey(@Nullable String repositoryKey, @Nullable String key) {
+    if (repositoryKey == null || key == null) {
+      return null;
+    }
+    return findByKey(RuleKey.of(repositoryKey, key));
+  }
+
+  @Override
+  @CheckForNull
+  public Rule findByKey(RuleKey key) {
+    return rulesByKey.get(key);
+  }
+
+  @Override
+  @CheckForNull
+  public Rule find(@Nullable RuleQuery query) {
+    if (query == null) {
+      return null;
+    }
+
+    return rulesByRuleDefinition.entrySet().stream()
+      .filter(entry -> matchQuery(entry.getKey(), query))
+      .sorted(FIND_BY_QUERY_ORDER)
+      .map(Map.Entry::getValue)
+      .findFirst()
+      .orElse(null);
+  }
+
+  @Override
+  public Collection<Rule> findAll(@Nullable RuleQuery query) {
+    if (query == null) {
+      return Collections.emptyList();
+    }
+    return rulesByRuleDefinition.entrySet().stream()
+      .filter(entry -> matchQuery(entry.getKey(), query))
+      .sorted(FIND_BY_QUERY_ORDER)
+      .map(Map.Entry::getValue)
+      .collect(MoreCollectors.toList());
+  }
+
+  private static boolean matchQuery(RuleDefinitionDto ruleDefinition, RuleQuery ruleQuery) {
+    if (RuleStatus.REMOVED.equals(ruleDefinition.getStatus())) {
+      return false;
+    }
+    String repositoryKey = ruleQuery.getRepositoryKey();
+    if (ruleQuery.getRepositoryKey() != null && !repositoryKey.equals(ruleDefinition.getRepositoryKey())) {
+      return false;
+    }
+    String key = ruleQuery.getKey();
+    if (key != null && !key.equals(ruleDefinition.getRuleKey())) {
+      return false;
+    }
+    String configKey = ruleQuery.getConfigKey();
+    return configKey == null || configKey.equals(ruleDefinition.getConfigKey());
+  }
+
+  private static Rule toRule(RuleDefinitionDto ruleDefinition, List<RuleParamDto> params) {
+    String severity = ruleDefinition.getSeverityString();
+    String description = ruleDefinition.getDescription();
+    RuleDto.Format descriptionFormat = ruleDefinition.getDescriptionFormat();
+
+    org.sonar.api.rules.Rule apiRule = new org.sonar.api.rules.Rule();
+    apiRule
+      .setName(ruleDefinition.getName())
+      .setLanguage(ruleDefinition.getLanguage())
+      .setKey(ruleDefinition.getRuleKey())
+      .setConfigKey(ruleDefinition.getConfigKey())
+      .setIsTemplate(ruleDefinition.isTemplate())
+      .setCreatedAt(new Date(ruleDefinition.getCreatedAt()))
+      .setUpdatedAt(new Date(ruleDefinition.getUpdatedAt()))
+      .setRepositoryKey(ruleDefinition.getRepositoryKey())
+      .setSeverity(severity != null ? RulePriority.valueOf(severity) : null)
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove this use of "RulePriority"; it is deprecated. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)